### PR TITLE
RxJava2 Migration: Return single/maybe where appropriate.

### PR DESCRIPTION
**Pull request by tairrzayev**
This attempts to deal with https://github.com/NYTimes/Store/issues/22

Most prominent changes:
- `Store` methods `get()` and `fetch()` will now return `Single`. 
- `Persister.read()` will now return `Maybe` because obviously it can complete without emissions if nothing was persisted before read.
 
This needs a good look at.

### CODE REVIEW

**Review by digitalbuddha — COMMENTED** [review]
Wow thank you so much for contribution!  
We recently merged #181 on the rx1 side which fixed a bug using the `networkBeforeStale` directive.  The crux of the issue was that Disk returning empty prevented proper propagation of network errors on first launch.  I'll can spend some time tomorrow adding the necesary test to your branch and finding a solution that hopefully works with the `Maybe` type.  Here is the new test if you want to offer a suggestion. https://github.com/NYTimes/Store/blob/develop/filesystem/src/test/java/com/nytimes/android/external/fs/StoreNetworkBeforeStaleFailTest.java#L48

**Comment by tairrzayev** [comment]
Thank you for heads up on that issue!

I went ahead and ported the original fix (https://github.com/NYTimes/Store/issues/181) into this branch in my latest commit - https://github.com/NYTimes/Store/pull/185/commits/20c0a2daf8864f247e7d57e81dbb617807ef01d7

**Comment by tairrzayev** [comment]
I've noticed a couple of possible optimisations / improvements which could be made to the `FSReader.read()` in due course (which I think is a topic for a separate issue / discussion):


```
String resolvedKey = pathResolver.resolve(key);
boolean exists = fileSystem.exists(resolvedKey);

if (exists) {
    try {
        BufferedSource bufferedSource = fileSystem.read(resolvedKey);
        emitter.onSuccess(bufferedSource);
        emitter.onComplete();
    } catch (FileNotFoundException e) {
        emitter.onError(e);
    }
} else {
    emitter.onError(new FileNotFoundException(ERROR_MESSAGE + resolvedKey));
}
```

- `exists()` call seems redundant as `FileSystem.read()` already declares `FileNotFoundException` being thrown and in fact, it does the same check & throws it if the file does not exist in `FSFile.source()`, which backs the `FileSystem.read()` implementation.
- Arguably `emitter.onError()` from `else` branch leaks `FSReader` implementation details by returning the full path (which is derived in that method) while all the user should know if that there is no file matching their key.

With these points in mind, I'd propose the following implementation:

```
String resolvedKey = pathResolver.resolve(key);
try {
    BufferedSource bufferedSource = fileSystem.read(resolvedKey);
    emitter.onSuccess(bufferedSource);
    emitter.onComplete();
} catch (FileNotFoundException e) {
    emitter.onError(new FileNotFoundException("No file matching the key :" + key));
}
```

**Comment by pavlospt** [comment]
@tairrzayev Sounds like a nice optimization and helps on the readability as well :)

**Comment by digitalbuddha** [comment]
Thank you for pulling in fix! I'm still a bit concerned about Persister returning a maybe rather than a single. From the store's perspective a Persister  should return either a value or an error.  My concern is if a client implements their own persister and returns empty from it NetworkBeforeStale will not function correctly for the same reason that FSReader needed the pulled in fix.  Could you kindly make the adjustment to have Persister return a Single so that we can make sure that any user implemented Persister works correctly.

File: store/src/main/java/com/nytimes/android/external/store2/base/impl/RealInternalStore.java [hunk 1]
```diff
@@ -171,24 +173,37 @@ private void initMemCache(MemoryPolicy memoryPolicy) {
...
 
-    Observable<Parsed> readDisk(@Nonnull final Key key) {
+    Maybe<Parsed> readDisk(@Nonnull final Key key) {
+        return readDisk(key, null);
+    }
+
+    Maybe<Parsed> readDisk(@Nonnull final Key key, final Throwable error) {
         return persister().read(key)
-                .onErrorResumeNext(new OnErrorResumeWithEmpty<Raw>())
+                .onErrorResumeNext(new Function<Throwable, MaybeSource<? extends Raw>>() {
```
Comment (digitalbuddha) [inline store/src/main/java/com/nytimes/android/external/store2/base/impl/RealInternalStore.java]: Here's the part we need to either have Persisters return an error when  empty or have some other operator handle the empty maybe case when a throwable is passed into read disk.

**Comment by tairrzayev** [comment]
Hmm, having looked back at the solution for this problem, I wonder if we could solve it as follows instead (keeping the `readDisk()` as-is, without adding additional `Throwable` param):

```
     @Nonnull
     Single<Parsed> response(@Nonnull final Key key) {
         return fetcher()
                 .fetch(key)
                 .flatMap(new Function<Raw, SingleSource<Parsed>>() {
                     @Override
                     public SingleSource<Parsed> apply(@NonNull Raw raw) {
                         return persister().write(key, raw)
                                 .flatMap(new Function<Boolean, SingleSource<Parsed>>() {
                                     @Override
                                     public SingleSource<Parsed> apply(@NonNull Boolean aBoolean) {
                                         return readDisk(key).toSingle();
                                     }
                                 });
                     }
                 })
                 .onErrorResumeNext(new Function<Throwable, SingleSource<? extends Parsed>>() {
                     @Override
                     public SingleSource<? extends Parsed> apply(@NonNull Throwable throwable) {
                         if (stalePolicy == StalePolicy.NETWORK_BEFORE_STALE) {
-                            return readDisk(key, throwable).toSingle();
+                            return readDisk(key)
+                                    .switchIfEmpty(Maybe.<Parsed>error(throwable))
+                                    .toSingle();
                         }
                         return Single.error(throwable);
                     }
                 })
                 .doOnSuccess(new Consumer<Parsed>() {
                     @Override
                     public void accept(@NonNull Parsed parsed) {
                         notifySubscribers(parsed);
                     }
                 })
                 .doAfterTerminate(new Action() {
                     @Override
                     public void run() {
                         inFlightRequests.invalidate(key);
                     }
                 })
                 .cache();
     }
```

In this case we'll switch to error if the cache is empty. It feels to me that reading from the cache should be a `Maybe` as it may or may not return the data - missing cached data should not really be an error.

**Comment by digitalbuddha** [comment]
That looks lovely! I like your solution much better. Could you kindly make the adjustment. I will review this today and have a few review it tomorrow. Thank you again for your contribution!

**Comment by tairrzayev** [comment]
Awesome :+1: - change applied.

**Review by ramonaharrison — APPROVED** [review]

**Review by alexio — APPROVED** [review]

**Review by digitalbuddha — APPROVED** [review]
Thank you for changes!

**Pull request merged**